### DISCUSSION TITLE
Can we make a TypeScript parser with this?

### DISCUSSION THREAD

**Discussion by trusktr** [eclipse-langium/langium#895] [General]
Looking to make a language that is a superset/subset of TypeScript syntax, but with a different type checking implementation, and not JS output. Would `langium` be suitable for making a TS parser?

**Comment by msujew** [comment]
We've build some TypeScript like languages in the past (closed-source) with Langium. While some language/parser features are not completely trivial to implement (such as template strings), Langium should be capable of dealing with that fairly well.

There will be a lot of work on the scoping/type system level you'll need to do, but I imagine that Langium will serve fairly well for this task. Someone created an [EBNF grammar of TypeScript](https://github.com/frenchy64/typescript-parser/blob/master/typescript.ebnf) a while back. It's not up to spec (since it's from 2014) but it should well as a starting point. I would also recommend to take a look at the [langium-lox](https://github.com/langium/langium-lox) repo, which shows how to implement scoping and type system support for a Langium based language.

**Reply by trusktr** [reply]
Thank you for the input! I didn't realize Langium was using Chevrotain. That's neat!

Some folks and I are interested in taking TypeScript syntax (but that's about all) and making custom type checking (maybe compatible? Maybe optionally not?) and custom output (f.e. native or wasm).

Is there a difference between Langium and what it does *after* parsing based on a parser generated from a Langium grammar (compared to using Chevrotain's runtime-JS grammar definitions)? Is it still LL(K)?

**Reply by msujew** [reply]
> Thank you for the input! I didn't realize Langium was using Chevrotain. That's neat!

I have written a whole [blog post about this topic](https://www.typefox.io/blog/allstar-lookahead) :)

> Is there a difference between Langium and what it does after parsing based on a parser generated from a Langium grammar (compared to using Chevrotain's runtime-JS grammar definitions)? Is it still LL(K)?

Depending on your configuration, Langium supports LL(*) lookahead actually (but Chevrotain does as well using a plugin), take a look at the blog post if you want to know more. Anyway, we can't go beyond what Chevrotain offers us in regards to parsing, so the input grammar will always need to be in LL form. So no left recursion.

Of course, Langium does a lot more than Chevrotain _afterwards_. The parser step is a relatively minor thing for Langium. We are more concerned about automatic AST generation, cross reference resolution, semantic validation, code generation and LSP support. Maybe you already had a chance to look at [our framework overview](https://langium.org/docs/langium-overview/).

**Reply by trusktr** [reply]
I'm reading the article, and it is interesting. Thanks for writing that! 

Indeed I saw Langium's overview, very enticing.

If we roll with Langium, would it be possible to write our grammar using Chevrotain syntax (with the plugins and things Langium adds I imagine) instead of generating from a grammar language?

**Reply by msujew** [reply]
Unfortunately not. The information that Langium is using to provide all of the advanced features (AST generation, cross reference resolution, LSP services, etc.) comes straight from the annotated grammar. There's no pure programmatic way of using Langium.

**Reply by luan-xiaokun** [reply]
> We've build some TypeScript like languages in the past (closed-source) with Langium. While some language/parser features are not completely trivial to implement (such as template strings), Langium should be capable of dealing with that fairly well.

I'm interested in how to resolve template-related ambiguities, since the language I'm working with also has this template ambiguity. For example, `A<B>(C)` can be parsed as a template function call, and can also be parsed as a nested comparison. Could you hint at how to achieve this (if possible)?

**Reply by msujew** [reply]
@luan-xiaokun I think this warrants it's own discussion in this forum, since it's a bit more complicated. Do you mind creating one? By the way with "template strings" I mean stuff like <code>\`label: ${value}\`</code>, which require some changes in the tokenizer.

**Reply by luan-xiaokun** [reply]
Sure, I just started a new discussion in #921
